src/probnum/randprocs/kernels/_kernel.py

```diff
@@ -273,82 +278,136 @@ def matrix(
+            ``+ batch_shape_1`` in "C-order".
 
         Raises
         ------
         ValueError
-            If the shapes of the inputs don't match the specification.
-
-        See Also
-        --------
-        __call__: Evaluate the kernel more flexibly.
-
-        Examples
-        --------
-        See documentation of class :class:`Kernel`.
+            If the shape of either input is not of the form ``batch_shape_0 +``
+            :attr:`input_shape`.
         """
+        x0 = self._preprocess_linop_input(x0, argname="x0")
 
-        x0 = np.asarray(x0)
-        x1 = x0 if x1 is None else np.asarray(x1)
+        if x1 is not None:
+            x1 = self._preprocess_linop_input(x1, argname="x1")
 
-        # Shape checking
-        errmsg = (
-            "`{argname}` must have shape `({batch_dim},) + input_shape` or "
-            f"`input_shape`, where `input_shape` is `{self.input_shape}`, but an array "
-            "with shape `{shape}` was given."
+        k_matrix_x0_x1 = self._evaluate_matrix(x0, x1)
+
+        assert isinstance(k_matrix_x0_x1, np.ndarray)
+        assert k_matrix_x0_x1.shape == (
+            self.output_size_0 * x0.shape[0],
+            self.output_size_1 * (x1.shape[0] if x1 is not None else x0.shape[0]),
         )
 
-        if not 0 <= x0.ndim - self._input_ndim <= 1:
-            raise ValueError(errmsg.format(argname="x0", batch_dim="M", shape=x0.shape))
+        return k_matrix_x0_x1
 
-        if not 0 <= x1.ndim - self._input_ndim <= 1:
-            raise ValueError(errmsg.format(argname="x1", batch_dim="N", shape=x1.shape))
+    def linop(
+        self,
+        x0: ArrayLike,
+        x1: Optional[ArrayLike] = None,
+    ) -> linops.LinearOperator:
+        r""":class:`~probnum.linops.LinearOperator` representing the pairwise
+        covariances of evaluations of :math:`f_0` and :math:`f_1` at the given input
+        points.
 
-        # Pairwise kernel evaluation
-        if x0.ndim > self._input_ndim and x1.ndim > self._input_ndim:
-            return self(x0[:, None], x1[None, :])
+        Representing the resulting covariance matrix as a matrix-free :class:`~probnum.\
+        linops.LinearOperator` is often more efficient than a representation as a
+        :class:`~numpy.ndarray`, both in terms of memory and computation time,
+        particularly when using iterative methods to solve the associated linear
+        systems.
 
-        return self(x0, x1)
+        For instance, covariance matrices induced by separable covariance functions
+        (e.g. tensor products of covariance functions or separable multi-output kernels)
+        can often be represented as :class:`~probnum.linops.KroneckerProduct`\ s of
+        smaller covariance matrices and frameworks like :mod:`pykeops<pykeops.numpy>`
+        can be used to implement efficient matrix-vector products with covariance
+        matrices without needing to construct the entire matrix in memory.
```

Comment:
Love the detailed docstring here why `.linop` exists and what it is used for.

Reply:
😁 🙏🏻